modules/zip/zip_reader.cpp

```diff
@@ -0,0 +1,114 @@
+/*************************************************************************/
+/*  zip_reader.cpp                                                       */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "zip_reader.h"
+
+#include "core/error/error_macros.h"
+#include "core/io/zip_io.h"
+
+Error ZIPReader::open(String path) {
+	if (f) {
+		close();
+	}
+
+	zlib_filefunc_def io = zipio_create_io_from_file(&f);
+	uzf = unzOpen2(path.utf8().get_data(), &io);
+	return uzf != NULL ? OK : FAILED;
+}
+
+Error ZIPReader::close() {
+	ERR_FAIL_COND_V_MSG(!f, FAILED, "ZIPReader cannot be closed because it is not open.");
+
+	return unzClose(uzf) == UNZ_OK ? OK : FAILED;
+}
+
+PackedStringArray ZIPReader::get_files() {
+	ERR_FAIL_COND_V_MSG(!f, PackedStringArray(), "ZIPReader must be opened before use.");
+
+	List<String> s;
+
+	if (unzGoToFirstFile(uzf) != UNZ_OK) {
+		return PackedStringArray();
+	}
+
+	do {
+		char filename[256];
```

Comment:
Are we sure about this limit ?

filename contains not only the name of the file, but also it path within the archive (e.g. `foo/bar/spam.txt`)

So the limit is fine on Windows where path are limited to 250 characters (but I think to recall there is special case in Windows where this might be longer), but on Linux/Mac 255 is the max size of file name (and entire path is at least 1024 bytes).

So I think it would be better to set a bigger buffer here

Typically doing something like
```suggestion
		char filename[256];  // Note filename is a path !
		int err = unzGetCurrentFileInfo64(uzf, NULL, filename, sizeof(filename), NULL, 0, NULL, 0);
		if (err == UNZ_OK) {
			s.push_back(filename);
		} else {
			// Assume filename buffer was too small
			unz_file_info64 file_info;
			int err = unzGetCurrentFileInfo64(uzf, &file_info, NULL, 0, NULL, 0, NULL, 0);
			char long_filename_buff  = memalloc(file_info.size_filename);
			int err = unzGetCurrentFileInfo64(uzf, NULL, long_filename_buff, sizeof(filename), NULL, 0, NULL, 0);
			s.push_back(long_filename_buff);
			memfree(long_filename_buff);
		}
```

Reply:
> So the limit is fine on Windows where path are limited to 250 characters

By default, Windows `MAX_PATH` is 260 (but it's possible to disable the limit on Windows 10+). Unicode UNC (`\\?\` prefixed) paths are limited to 32K UTF-16 characters.

Reply:
I'm not sure what limits Zip itself imposes, but:

1. If you use `\\?\`-format paths or enable Win32 long paths via group policy or the registry on Windows 10, then you're subject to much longer filesystem-based length limits on Windows.
2. Various POSIX-specified libc functions are limited to `PATH_MAX` (`limits.h`) by a buffer allocation, which is typically 4096, but you [can work around that](https://insanecoding.blogspot.com/2007/11/pathmax-simply-isnt.html).

Here are the limits I'm aware of:

* **FAT32:** 32,760 "Unicode characters" for paths, 255 "Unicode characters" for filenames
* **exFAT:** 32,760 "Unicode characters", 255 "Unicode characters" for filenames
* **NTFS:** 32768 "Unicode characters" , 255 "Unicode characters" for filenames
* **ReFS:** 32768 "Unicode characters", 255 "Unicode characters" for filenames
* **UDF:** 1023 bytes, , 255 bytes for filenames
* **ext2/3/4:** No limit for paths (As I demonstrate with some Python [here](https://askubuntu.com/questions/859945/what-is-the-maximum-length-of-a-file-path-in-ubuntu/1169977#1169977)), 255 bytes for filenames
* **HFS+:** No limit for paths as far as I was able to google up, 255 bytes for filenames

As far as I'm aware, "Unicode characters" means 16-bit code units.

Reply:
ZIP format uses a 16-bit value to store filename length, so it's probably limited to 64K (UTF-8). Not sure if there are extra limitations, but it's probably better to always use `unz_file_info64` and allocate space for filename on demand.

Reply:
Also...

> So the limit is fine on Windows where path are limited to 250 characters

Actually, if I remember correctly, the limit is **260** characters and they chose that limit based on `A:\<some buffer that allocates for 256 chars><NUL>`.

> ZIP format uses a 16-bit value to store filename length, so it's probably limited to 64K (UTF-8).

Yeah. No clue about the upper limit, but there are plenty of google results with people complaining about this or that unzipping tool erroring out because the Zip file contains paths longer than 260 and the tool isn't using `\\?\`.